# Week One Homework

### While AI-assisted tools such as ChatGPT will be used in the completion of this assignment, the use of published documentation online will be used in the assistance of writing code.
<p>The documentation that will be used is as follows:<br>
    &emsp;
        - <a href="https://docs.python.org/3/">docs.python.org</a><br>
    &emsp;
        - <a href="https://pandas.pydata.org/docs/user_guide/index.html">pandas.pydata.org</a>
</p>
<p>The ChatGPT chat log:<br>
   &emsp;- <a href="https://chatgpt.com/share/da60cd77-3e9a-4c30-9dcf-901ed0570be7">https://chatgpt.com/share/da60cd77-3e9a-4c30-9dcf-901ed0570be7</a>
</p>

<h2>Pre-lecture Homework</h2>

<h3>Step 1: Grabbing a dataset from a tutorial and the subsequent relevant code to accompany it</h3>
<p>I will be using the <a href="https://raw.githubusercontent.com/datasets/co2-fossil-global/master/global.csv">"Global CO2 Emissions"</a> Dataset from my TUT0207 tutorial. The accompanying code is displayed below.

In [58]:
import pandas as pd

# Step 1: Download the CSV file
url = 'https://raw.githubusercontent.com/datasets/co2-fossil-global/master/global.csv'

# Step 3: Read the CSV data into a DataFrame
df = pd.read_csv(url)

# Step 4: Display the DataFrame (optional)
print(df.head())

   Year  Total  Gas Fuel  Liquid Fuel  Solid Fuel  Cement  Gas Flaring   
0  1751      3         0            0           3       0            0  \
1  1752      3         0            0           3       0            0   
2  1753      3         0            0           3       0            0   
3  1754      3         0            0           3       0            0   
4  1755      3         0            0           3       0            0   

   Per Capita  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  


<h3>Step 2: Asking ChatGPT to produce code on how many rows of data are contained within the dataset</h3>
<p>ChatGPT has given me a recommendation to use the "df.shape" attribute, as it is an array containing the length and width of the dataframe. By asking for the zero index, we are retrieving only the number of rows whereas the one index would return the number of collumns.</p>

In [59]:
# Print the number of rows in the DataFrame
print(f'The DataFrame has {df.shape[0]} rows.')

The DataFrame has 260 rows.


<h3>Step 3: Asking ChatGPT to produce code which will create summaries about the columns for my dataframe</h3>

<p>The code seen below is taken directly from ChatGPT and WILL PRODUCE an error.</p>

In [60]:
# 1. Get summary statistics for numerical columns
summary_stats = df.describe()
print("Summary statistics:")
print(summary_stats)

# 2. Get value counts for a specific column (replace 'column_name' with the actual column)
column_name = 'column_name'
value_counts = df[column_name].value_counts()
print(f"\nValue counts for '{column_name}':")
print(value_counts)

Summary statistics:
              Year        Total    Gas Fuel  Liquid Fuel   Solid Fuel   
count   260.000000   260.000000   260.00000   260.000000   260.000000  \
mean   1880.500000  1402.788462   185.20000   495.819231   674.569231   
std      75.199734  2253.098527   396.58556   934.308074   868.368580   
min    1751.000000     3.000000     0.00000     0.000000     3.000000   
25%    1815.750000    12.750000     0.00000     0.000000    12.750000   
50%    1880.500000   239.500000     0.00000     3.000000   236.000000   
75%    1945.250000  1385.000000    59.50000   279.250000  1023.500000   
max    2010.000000  9167.000000  1702.00000  3122.000000  3842.000000   

           Cement  Gas Flaring  Per Capita  
count  260.000000   260.000000   61.000000  
mean    34.161538    13.065385    1.054754  
std     78.899604    26.311315    0.178630  
min      0.000000     0.000000    0.640000  
25%      0.000000     0.000000    0.940000  
50%      0.000000     0.000000    1.120000  
75%    

KeyError: 'column_name'

<p>To fix this issue, we look at the interpreter's throwing of the KeyError and realize that we need a more specific solution for each individual colunmn within this code. I will be referencing the Pandas documentation and python documentation to debug this code rather than trying to describe my dataset to ChatGPT.</p>

In [61]:
# 1. Get summary statistics for numerical columns
summary_stats = df.describe()
print("Summary statistics:")
print(summary_stats)

# 2. Get value counts for a specific column (replace 'column_name' with the actual column) (INCORRECT)

## The code below was given directly by ChatGPT and caused bugs before.
# column_name = 'column_name'

## https://www.geeksforgeeks.org/ was also used in trying to remedy this bug.
# 2. Get the values of each element in the column names by using df.columns

column_names = list(df.columns)

# 3. Use a for-each loop to provide summaries for each individual column present in the dataframe

for column_name in column_names:
    value_counts = df[column_name].value_counts()
    print(f"\nValue counts for '{column_name}':")
    print(value_counts)

Summary statistics:
              Year        Total    Gas Fuel  Liquid Fuel   Solid Fuel   
count   260.000000   260.000000   260.00000   260.000000   260.000000  \
mean   1880.500000  1402.788462   185.20000   495.819231   674.569231   
std      75.199734  2253.098527   396.58556   934.308074   868.368580   
min    1751.000000     3.000000     0.00000     0.000000     3.000000   
25%    1815.750000    12.750000     0.00000     0.000000    12.750000   
50%    1880.500000   239.500000     0.00000     3.000000   236.000000   
75%    1945.250000  1385.000000    59.50000   279.250000  1023.500000   
max    2010.000000  9167.000000  1702.00000  3122.000000  3842.000000   

           Cement  Gas Flaring  Per Capita  
count  260.000000   260.000000   61.000000  
mean    34.161538    13.065385    1.054754  
std     78.899604    26.311315    0.178630  
min      0.000000     0.000000    0.640000  
25%      0.000000     0.000000    0.940000  
50%      0.000000     0.000000    1.120000  
75%    

<h3>Step 4: Describing descrepancies between the summaries of each column and that of the dataframe</h3>
<p>When analyzing this dataframe, it is clear that for certain year, somee values may not have been recorded. However, missing attributes in certain columns will not relfect or impact the size of the dataframe as a whole, as the size of the 2D array is dependent on the maximum number of rows for any column rather than the maximum number of rows that they share. Thus, this descrepancy is easily explained and is to be expected.</p><br>

<h3>Step 5: Using ChatGPT to describe the differences between attributes and methods</h3>
<p>Using the paradigm of Object Oriented Programming (which Python heavily implements), we can easily explain the differences between the methods which introduce parentheses at the end, and the attributes (or instance/static variables) that are called upon without the use of parentheses.</p>
<p></p>
<p><b>For example, take a look at the code below:</b></p>

In [62]:
class dataframe:
    
    def __init__(self, rows: int, columns: int):
        self.rows = rows
        self.columns = columns
        self.shape = [rows, columns]
        
    def describe(self):
        return f"The dimensions of the dataframe consists of {self.rows} rows and {self.columns} columns."
    

<p>In the simple sample class written above, we can see that self.columns refers to a variable that belongs to this instance of the dataframe. If the user creates a new dataframe "df" and gives it the dimensions of 50 rows and 50 columns, the value of df.shape would also be equal to [50, 50].</p>
<p></p>
<p>If the user was to then make another dataframe, this time called "df_2", and assigns it the dimensions of 60 rows and 70 columns, the value of df_2.shape would then be equal to [60, 70].</p>
<p></p>
<p><b>Notice how the creation of df_2 does not modify the value of the original df.shape- and thus is the unique difference between an attribute and a regular variable.</b></p>

<p>However, a method is a way for any user to access or modify the values within an object or class. Thus, we can also refer to these methods as a form of "interface".</p>
<p></p>
<p><b>For example, let's examine the code below:</b></p>

In [63]:
if __name__ == "__main__":
    df = dataframe(rows=80, columns=100)
    print(df.describe())

The dimensions of the dataframe consists of 80 rows and 100 columns.


<p>The use of df.describe() calls upon the dataframe.describe() to help us access information about the object (in this case, our dataframe) and thus fulfills the requirements for a method. We call this type of method an "accessor method".</p>

## Post-lecture Homework

### Step 6: Giving definitions of the output values from `df.describe()`
<p><i>Please note: instead of using ChatGPT, this step was completed using source materials from the Pandas Documentation, which can be found <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html#pandas.DataFrame.describe">here.</a></i></p>

<p><b>df.count</b><br>
    When using df.describe(), the value "count" refers to the number of non-null (or, non-None type) values found within a given dataframe.</p>
<p><b>df.mean</b><br>
    The variable df.mean which gets outputting as a result of the method df.describe() gives the arithmetic mean of the values in the dataframe.</p>
<p><b>df.std</b><br>
    The df.describe() std value outputs the standard deviation of the dataset.</p>
<p><b>df.min</b><br>
    The min value describes the minimum value for each value in the dataframe.</p>
<p><b>df.percentiles (25%, 50%, 75%)</b><br>
    The values of 25%, 50%, and 75% refer to the value of the percentiles. For instance, 25% refers to the value that would correspond with the maximum value for the bottom 25% of values in the dataset. 75% thus also refers to the minimum value for the top 25% of values within the dataset. 50% refers to the median.</p>
<p><b>df.max</b><br>
    The max value refers to the maximum value for all values within the dataframe.</p>


### Step 7: Describing the differences between `df.dropna()` and `del df['col']`
<p><b>&emsp;1. Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']</b><br>
    &emsp;&emsp;In this case, when a whole row consists of missing values, it would be wiser to use df.dropna() to remove the least amount of useful information.</p>
<p><b>&emsp;2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()</b><br>
    &emsp;&emsp;If there consists of a column with mostly null values with rows with few null values, then it makes more sense to remove the entire column using del <br>&emsp;&emsp;df['col'] instead.</p>
<p><b>&emsp;3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important</b><br>
    &emsp;&emsp;Suppose that our dataframe has a size of 25 rows and 5 columns. If one column has only 5 non-null values, then the use of df.dropna() first would <br>&emsp;&emsp;remove 20 rows of useful information. However, when del df['col'] is applied first, minimal data is lost as the use of df.dropna() would remove zero rows. </p>
<p><b>&emsp;4. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give <br>&emsp;&emsp;a justification for your approach, including a "before and after" report of the results of your approach for your dataset.</b><br>
    &emsp;&emsp;<i>See code below.</i></p>

In [64]:
url = 'https://raw.githubusercontent.com/datasets/co2-fossil-global/master/global.csv'

df = pd.read_csv(url)
print(df.head())

   Year  Total  Gas Fuel  Liquid Fuel  Solid Fuel  Cement  Gas Flaring   
0  1751      3         0            0           3       0            0  \
1  1752      3         0            0           3       0            0   
2  1753      3         0            0           3       0            0   
3  1754      3         0            0           3       0            0   
4  1755      3         0            0           3       0            0   

   Per Capita  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  


In [65]:
print(df.describe(include='all'))

              Year        Total    Gas Fuel  Liquid Fuel   Solid Fuel   
count   260.000000   260.000000   260.00000   260.000000   260.000000  \
mean   1880.500000  1402.788462   185.20000   495.819231   674.569231   
std      75.199734  2253.098527   396.58556   934.308074   868.368580   
min    1751.000000     3.000000     0.00000     0.000000     3.000000   
25%    1815.750000    12.750000     0.00000     0.000000    12.750000   
50%    1880.500000   239.500000     0.00000     3.000000   236.000000   
75%    1945.250000  1385.000000    59.50000   279.250000  1023.500000   
max    2010.000000  9167.000000  1702.00000  3122.000000  3842.000000   

           Cement  Gas Flaring  Per Capita  
count  260.000000   260.000000   61.000000  
mean    34.161538    13.065385    1.054754  
std     78.899604    26.311315    0.178630  
min      0.000000     0.000000    0.640000  
25%      0.000000     0.000000    0.940000  
50%      0.000000     0.000000    1.120000  
75%     12.000000     0.000

In [66]:
df_dropped = df
del df_dropped['Per Capita']

for column in df_dropped:
    for i in range(len(df_dropped['Year'])):
        value = df_dropped[column][i]
        if value == 0:
            df_dropped[column][i] = None

df_dropped.dropna(inplace=True)
print(df_dropped.head())

     Year  Total  Gas Fuel  Liquid Fuel  Solid Fuel  Cement  Gas Flaring
199  1950   1630      97.0        423.0        1070    18.0         23.0
200  1951   1767     115.0        479.0        1129    20.0         24.0
201  1952   1795     124.0        504.0        1119    22.0         26.0
202  1953   1841     131.0        533.0        1125    24.0         27.0
203  1954   1865     138.0        557.0        1116    27.0         27.0


C:\Users\steve\AppData\Local\Temp\ipykernel_17296\2656622103.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped[column][i] = None
C:\Users\steve\AppData\Local\Temp\ipykernel_17296\2656622103.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped[column][i] = None
C:\Users\steve\AppData\Local\Temp\ipykernel_17296\2656622103.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped[column][i] = None


In [67]:
print(df_dropped.describe())

              Year        Total     Gas Fuel  Liquid Fuel   Solid Fuel   
count    61.000000    61.000000    61.000000    61.000000    61.000000  \
mean   1980.000000  4984.114754   769.327869  2005.229508  2012.081967   
std      17.752934  2070.838720   474.350599   850.674054   696.221907   
min    1950.000000  1630.000000    97.000000   423.000000  1070.000000   
25%    1965.000000  3130.000000   337.000000  1219.000000  1448.000000   
50%    1980.000000  5113.000000   740.000000  2289.000000  1921.000000   
75%    1995.000000  6398.000000  1157.000000  2605.000000  2414.000000   
max    2010.000000  9167.000000  1702.000000  3122.000000  3842.000000   

           Cement  Gas Flaring  
count   61.000000    61.000000  
mean   141.901639    55.688525  
std    106.848289    23.983009  
min     18.000000    23.000000  
25%     59.000000    37.000000  
50%    120.000000    48.000000  
75%    197.000000    66.000000  
max    450.000000   110.000000  


<p>Due to the lack of null values in the dataset, I just set all zero values to null values to simulate the existence of null values for the completion of this assignment.<br><br>Using df.describe(), the result is the same but the dataset has drastically changed. In the latter dataset, the indices start from 199 instead of zero due to the values omitted from the first dataset. This reduces the majority of the 18th and 19th centuries, which had empty values for the vast majority of energy types.</p>

### Step 8: Brief explanations of the questions below

<p><i>For this question, for any mentions to a search engine, the DuckDuckGo search engine was used to test the scenarios provided below.</i></p>

<p><b>1. Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you</b><br>
    The df.groupby() method re-arranges the dataframe based on a certain element in the dataset. The "col1" call would organize the dataframe based on the elements in the column labeled "col1", then it would consider the values for the elements in "col2". Below is some code to demonstrate the idea.</p>

In [68]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

df = pd.read_csv(url)
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class   
0         0       3    male  22.0      1      0   7.2500        S  Third  \
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [69]:
df.groupby('sex')['survived'].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

<p>For the code above, the use of groupby('sex')['survived'] allows us to analyze the survival rates of males and females. First, calling 'sex' as a parameter of groupby allows us to separate the dataset into two distinct groups- males and females. Then, by specifically analyzing the second element of "survived" (which acts as a pseudo-boolean through the discreet values of 0 and 1), this allows us to see the survival rates using the mean. For females, the mean is 0.742, which represents a survival rate of 74.2% for all females. For males, the mean is 0.189, which represents a surival rate of 18.9%.</p>

<p><b>2. Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?</b><br>
The main difference is that df.describe() describes the number of non-missing values for each column throughout the dataset. However, when calling df.groupby() before calling describe(), the count refers to the number of non-missing values for each group that is created by calling groupby().</p>

<p><b>3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT</b></p>

<p><b>A. Forget to include import pandas as pd in your code</b><br>
For this first error, both the ChatGPT and the web search (which returned a StackOverflow forum result) gave a simple one-line solution that helped solve this simple error. However, without needing to wait for ChatGPT's response generation, the web search was considerably faster and simpler.</p>

<p><b>B. Mistype "titanic.csv" as "titanics.csv"</b><br>
For this second error, the online search suggests to check the scope and path of the file, and that if the file exists outside the scope of the current project then the absolute path is required. ChatGPT, on the other hand, gave multiple possibilities for the cause of this error- including the possibility of a typo and how to remedy the problem by replacing 'titanics.csv' with 'titanic.csv'. While outside the scope of this assignment, it would be remiss to not mention that many high-level IDEs (such as PyCharm and Visual Studio Code) both include spellcheck functionalities which can point out the potential error even before runtime.</p>

<p><b>C. Try to use a dataframe before it's been assigned into the variable</b><br>
Using both ChatGPT and a web search (which of course, then leads to a StackOverflow forum) yield similar results. ChatGPT gave a more thoughtful response, considering the possibility of case-sensitivity. However, the forum post was much more consise and addressed the variable assignment error head on and thus led to a more efficient workflow. Again, a powerful IDE will also catch this issue before runtime and makes this simple mistake easy to spot.</p>

<p><b>D. Forget one of the parentheses somewhere the code</b><br>
In this case, both ChatGPT and a web search(which led to an article instead of a forum post this time) gave a simple answer and explained this silly syntax error that would have also been caught by a powerful IDE. In this case, using ChatGPT became remarkably slower due to the time needed to generate answers as compared to loading a simple webpage.</p>

<p><b>E. Mistype one of the names of the chained functions with the code</b>
Interestingly enough, it seems that ChatGPT seems to consider aspects Occam's razor for its responses- ChatGPT prefers the simplist solutions and infers that the prompt author is new to programming. However, in forum posts from StackOverflow, it seems that the users seem to assume that the users of the site are already fairly proficient at the language, and thus will try recommending more complex possibilities- such as the existance of whitespace characters within the syntax of the code. However, most surprisingly, the most helpful tool was neither the web search or AI-assisted debugging- but rather, the error message from a professional and programmer-first oriented IDE. For example, my IDE of choice, when faced with this simple problem, produced the following error message:</p>

#### `AttributeError: 'DataFrame' object has no attribute 'hea'. Did you mean: 'head'?`

<p>The IDE will even give suggestions for the methods available, negating the need for ChatGPT or StackOverflow, and thus should be considered when determining which tool is most helpful for debugging.</p>

<p><b>F. Use a column name that's not in your data for the groupby and column selection</b><br>
Both ChatGPT and a web search gave similar results that describes either a typo or checking to see if the index existed in the dataset. In this case, both of these gave easy to follow steps that describe the simple error and can help solve the problem easily.</p>

<p><b>G. Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question</b><br>
Due to the simple nameerror that was thrown, the web search attempts to solve the problem by treating it as any regular variable. However, ChatGPT, understanding the context due to the history of the chat, was able to deduce the problem and immediately suggested the use of single quotes to solve the problem. In this case, the use of AI was able to solve this problem much easier than the use of traditional web search debugging.</p>

<p><i>Overall, sometimes the use of AI can be a better debugging tool than traditional web searches. However, the power of AI breaks down when using niche libraries as ChatGPT cannot understand the nuances of each specific external library of python. Not to mention, as ChatGPT does not have its own Python interpreter built in, it has no way of telling whether the code it's generating will actually run without any errors or whether the logic works. For this reason, I still prefer to use traditional web search debugging and consulting documentation when in doubt.</i></p>

### Step 9: Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

#### Yes. I have.